components/trace/appmap/classmap/entity.mjs

```diff
@@ -91,24 +92,29 @@ export const getEntityLabelArray = (entity) =>
 export const getEntityQualifiedName = (entity, maybe_parent_entity) => {
   if (entity.type === "class") {
     return entity.name;
-  } else if (entity.type === "function") {
-    assert(
-      maybe_parent_entity !== null,
-      "function entity at root level",
-      InternalAppmapError,
-    );
```

Comment:
If `wrapRootEntityArray` is responsible for ensuring functions have a parent, I'd assume this assertion is also made there. Enforcing an assertion solely downstream causes the root issue to become opaque and difficult to debug once the assertion is broken. 

I'm not really sure _why_ this is happening, but this assertion is what's causing the agent to panic. The resulting AppMap is valid, so from an end-user experience, what's best? I'd argue that a clean exit code and the getting the resulting data out is a much better experience than a panic. Most users won't take the time to report the issue, they'll just move on.

Reply:
If you look at the content of `wrapRootEntityArray`, you will see that this assertion will always be verified. That that there is is impossible the invoke this function and break this assertion. I only use assertions when the blame can be put on the consumer of the method. Otherwise you pretty much fill up your code with assertions. But this is some general philosophical mombo-jumbo.

Of course it is better to produce a valid appmap than crashing. I usually only add assertions when subsequent operations do not make sense if the assertion is failing. In this case, I added the assertion because the parent code entity was needed to compute the qualified name of the function.

Anyhow, I found the root cause of the issue: wrapping root-level functions was done *before* exclusions. https://github.com/getappmap/appmap-agent-js/pull/185